assets/src/apps/delivery/components/ActivityRenderer.tsx

```diff
@@ -80,8 +82,39 @@ const ActivityRenderer: React.FC<ActivityRendererProps> = ({
   onRequestLatestState = async () => ({ snapshot: {} }),
 }) => {
   const isPreviewMode = useSelector(selectPreviewMode);
-  const currentUserId = 1; // TODO from state
+  const currentUserId = useSelector(selectUserId);
 
+  const saveUserData = async (attemptGuid: string, partAttemptGuid: string, payload: any) => {
+    const objId = `${payload.key}`;
+    debouncedSaveData({ isPreviewMode, payload, objId, value: payload.value });
+  };
+
+  const readUserData = async (attemptGuid: string, partAttemptGuid: string, payload: any) => {
+    // Read only the key from the simid
+    const objId = `${payload.key}`;
+    const data = debouncedReadData({ isPreviewMode, payload, objId });
+    return data;
+  };
+
+  const debouncedReadData = debounce(
+    async ({ isPreviewMode, payload, objId }) => {
+      const retrievedData = await Extrinsic.readGlobalUserState(
+        [`${payload.simId}`],
+        isPreviewMode,
+      );
+      return JSON.parse(retrievedData)[payload.simId]?.[objId];
```

Comment:
this also throws an exception when there is no data, however I also think that any parsing that needs to be done should be in the extrinsic method instead

Reply:
Addressed.